--- net/minecraft/src/World.java
+++ net/minecraft/src/World.java
@@ -276,6 +276,7 @@
 	}
 
 	public void saveWorld(boolean var1, IProgressUpdate var2) {
+	    if(!RMCSHNative.ShouldSave) return;
 		if(this.chunkProvider.canSave()) {
 			if(var2 != null) {
 				var2.func_594_b("Saving level");
@@ -291,12 +292,14 @@
 	}
 
 	private void saveLevel() {
+	    if(!RMCSHNative.ShouldSave) return;
 		this.checkSessionLock();
 		this.saveHandler.saveWorldInfoAndPlayer(this.worldInfo, this.playerEntities);
 		this.field_28108_z.saveAllData();
 	}
 
 	public boolean func_650_a(int var1) {
+	    if(!RMCSHNative.ShouldSave) return false;
 		if(!this.chunkProvider.canSave()) {
 			return true;
 		} else {
--- net/minecraft/src/EntityRenderer.java
+++ net/minecraft/src/EntityRenderer.java
@@ -10,6 +10,7 @@
 import org.lwjgl.opengl.GLContext;
 import org.lwjgl.opengl.NVFogDistance;
 import org.lwjgl.util.glu.GLU;
+import org.lwjgl.input.Keyboard;
 
 public class EntityRenderer {
 	public static boolean field_28135_a = false;
@@ -371,7 +372,11 @@
 			}
 
 			long var8;
+			RMCSHNative.Update();
 			if(this.mc.theWorld != null) {
+				if(!RMCSHNative.Loaded) {
+					RMCSHNative.Setup(this.mc);
+				}
 				if(this.mc.gameSettings.limitFramerate == 0) {
 					this.renderWorld(var1, 0L);
 				} else {
@@ -393,6 +398,9 @@
 				if(!this.mc.gameSettings.hideGUI || this.mc.currentScreen != null) {
 					this.mc.ingameGUI.renderGameOverlay(var1, this.mc.currentScreen != null, var16, var17);
 				}
+				if(!Keyboard.isKeyDown(Keyboard.KEY_F3)) {
+					RMCSHNative.Draw();
+				}
 			} else {
 				GL11.glViewport(0, 0, this.mc.displayWidth, this.mc.displayHeight);
 				GL11.glMatrixMode(GL11.GL_PROJECTION);
@@ -400,6 +408,7 @@
 				GL11.glMatrixMode(GL11.GL_MODELVIEW);
 				GL11.glLoadIdentity();
 				this.func_905_b();
+				RMCSHNative.Reset();
 				if(this.mc.gameSettings.limitFramerate == 2) {
 					var8 = (this.field_28133_I + (long)(1000000000 / var7) - System.nanoTime()) / 1000000L;
 					if(var8 < 0L) {
